src/core/reactor.cc

```diff
@@ -1651,6 +1651,13 @@ reactor::posix_listen(socket_address sa, listen_options opts) {
 
     try {
         fd.bind(sa.u.sa, sa.length());
+
+        if (sa.is_af_unix() && opts.unix_domain_socket_permissions) {
+            auto mode = static_cast<mode_t>(*opts.unix_domain_socket_permissions);
+            auto syscall_result = wrap_syscall<int>(::chmod(sa.u.un.sun_path, mode));
+            syscall_result.throw_if_error();
```

Comment:
Should unlink on failure.

Add a comment explaining why ::chmod() doesn't cause a stall.

Reply:
I found out about `fchmod` system call. Here it should be more suitable than `chmod` since we have a file descriptor for the socket. I will change the code to use `fchmod` and then unlinking on failure won't be necessary (`fchmod` will be called before `bind`, so there is nothing to unlink).

Reply:
Very nice, I didn't know that fchmod **before** the bind was even possible, but I see multiple references to this trick online. This nicely solves the race condition of bind-and-then-chmod. I didn't know this works, and apparently it's a relatively new Linux trick - on non-Linux Unix-like systems you'll get an EINVAL (POSIX https://pubs.opengroup.org/onlinepubs/009696699/functions/fchmod.html says it's "unspecified"). 

Please note that umask still applies (I believe), so whatever permissions you'll try to set, actually 022 will be removed from it (your umask).

Reply:
fchmod does work, but it *also* removes the umask bits.
The question is why is it bad? open(2) and creat(2) also lets you choose the "mode" bits on file creation - but neither of them takes your given mode as-is, both subtract umask from it. So it is expected by users.

Note that if you use chmod() after the bind to just **increase** permissions, there is no security risk. So if you really need to override the umask and readd these bits, you can. I just don't think anybody needs to do this. What do you plan to use this feature for?

Finally, please note that umask() is shared by all threads of a process. So changing it is a rather nasty thing to do, and should be done by the application once on initialization - not before some bind() call. I wasn't suggesting that you should change umask in each call - I was thinking more along the line of accepting the current umask(), or asking the user to change it once in their application.